core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -587,140 +482,143 @@ fd_table_detach(struct fd_table *ft, __wasi_fd_t fd, struct fd_object **fo)
-        *rights_base = RIGHTS_SOCKET_BASE;
-        *rights_inheriting = RIGHTS_SOCKET_INHERITING;
-    }
-    else {
-        return __WASI_EINVAL;
+    struct __wasi_filestat_t buf;
+    __wasi_errno_t error = os_fstat(fd, &buf);
+
+    if (error != __WASI_ESUCCESS)
+        return error;
+
+    *type = buf.st_filetype;
+
+    switch (buf.st_filetype) {
+        case __WASI_FILETYPE_BLOCK_DEVICE:
+            *rights_base = RIGHTS_BLOCK_DEVICE_BASE;
+            *rights_inheriting = RIGHTS_BLOCK_DEVICE_INHERITING;
+            break;
+        case __WASI_FILETYPE_CHARACTER_DEVICE:
+            error = os_isatty(fd);
+
+            if (error == __WASI_ESUCCESS) {
+                *rights_base = RIGHTS_TTY_BASE;
+                *rights_inheriting = RIGHTS_TTY_INHERITING;
+            }
+            else {
+                *rights_base = RIGHTS_CHARACTER_DEVICE_BASE;
+                *rights_inheriting = RIGHTS_CHARACTER_DEVICE_INHERITING;
+            }
+            break;
+        case __WASI_FILETYPE_DIRECTORY:
+            *rights_base = RIGHTS_DIRECTORY_BASE;
+            *rights_inheriting = RIGHTS_DIRECTORY_INHERITING;
+            break;
+        case __WASI_FILETYPE_REGULAR_FILE:
+            *rights_base = RIGHTS_REGULAR_FILE_BASE;
+            *rights_inheriting = RIGHTS_REGULAR_FILE_INHERITING;
+            break;
+        case __WASI_FILETYPE_SOCKET_DGRAM:
+        case __WASI_FILETYPE_SOCKET_STREAM:
+            *rights_base = RIGHTS_SOCKET_BASE;
+            *rights_inheriting = RIGHTS_SOCKET_INHERITING;
+            break;
+        case __WASI_FILETYPE_SYMBOLIC_LINK:
+        case __WASI_FILETYPE_UNKNOWN:
+            // If we don't know the type, allow for the maximum set of
+            // rights
+            *rights_base = RIGHTS_ALL;
+            *rights_inheriting = RIGHTS_ALL;
+            break;
+        default:
+            return __WASI_EINVAL;
     }
 
+    wasi_libc_file_access_mode access_mode;
+    error = os_handle_get_access_mode(fd, &access_mode);
+
+    if (error != __WASI_ESUCCESS)
+        return error;
+
     // Strip off read/write bits based on the access mode.
-    switch (fcntl(fd, F_GETFL) & O_ACCMODE) {
-        case O_RDONLY:
+    switch (access_mode) {
+        case WASI_LIBC_ACCESS_MODE_READ_ONLY:
             *rights_base &= ~(__wasi_rights_t)__WASI_RIGHT_FD_WRITE;
             break;
-        case O_WRONLY:
+        case WASI_LIBC_ACCESS_MODE_WRITE_ONLY:
             *rights_base &= ~(__wasi_rights_t)__WASI_RIGHT_FD_READ;
             break;
     }
-    return 0;
-#endif
+
+    return error;
 }
 
-// Returns the underlying file descriptor number of a file descriptor
+// Returns the underlying handle of a file descriptor
 // object. This function can only be applied to objects that have an
-// underlying file descriptor number.
-static int
-fd_number(const struct fd_object *fo)
+// underlying os handle associated.
+static os_handle
+fd_get_handle(const struct fd_object *fo)
```

Comment:
given that we no longer do the validation, and the function doesn't provide any abstraction (since `os_handle` is public), I don't know whether there's much value in keeping it - but up to you.

Reply:
Yup, we can get rid of it 👍